packages/finance/src/vesting/vesting.cairo

```diff
@@ -0,0 +1,210 @@
+            self: @ComponentState<TContractState>,
+            token: ContractAddress,
+            total_allocation: u256,
+            timestamp: u64,
+            start: u64,
+            duration: u64,
+            cliff: u64,
+        ) -> u256;
+    }
+
+    #[embeddable_as(VestingImpl)]
+    impl Vesting<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl Ownable: OwnableComponent::HasComponent<TContractState>,
+        +VestingScheduleTrait<TContractState>
+    > of interface::IVesting<ComponentState<TContractState>> {
+        /// Returns the timestamp marking the beginning of the vesting period.
+        fn start(self: @ComponentState<TContractState>) -> u64 {
+            self.Vesting_start.read()
+        }
+
+        /// Returns the timestamp marking the end of the cliff period.
+        fn cliff(self: @ComponentState<TContractState>) -> u64 {
+            self.Vesting_cliff.read()
+        }
+
+        /// Returns the total duration of the vesting period.
+        fn duration(self: @ComponentState<TContractState>) -> u64 {
+            self.Vesting_duration.read()
+        }
+
+        /// Returns the timestamp marking the end of the vesting period.
+        fn end(self: @ComponentState<TContractState>) -> u64 {
+            self.start() + self.duration()
+        }
+
+        /// Returns the already released amount for a given `token`.
+        fn released(self: @ComponentState<TContractState>, token: ContractAddress) -> u256 {
+            self.Vesting_released.read(token)
+        }
+
+        /// Returns the amount of a given `token` that can be released at the time of the call.
+        fn releasable(self: @ComponentState<TContractState>, token: ContractAddress) -> u256 {
+            let now = starknet::get_block_timestamp();
+            let vested_amount = self.resolve_vested_amount(token, now);
+            let released_amount = self.released(token);
+            if vested_amount >= released_amount {
+                vested_amount - released_amount
+            } else {
+                0
+            }
+        }
+
+        /// Returns the total vested amount of a specified `token` at a given `timestamp`.
+        fn vested_amount(
+            self: @ComponentState<TContractState>, token: ContractAddress, timestamp: u64
+        ) -> u256 {
+            self.resolve_vested_amount(token, timestamp)
+        }
+
+        /// Releases the amount of a given `token` that has already vested.
+        ///
+        /// Emits an `AmountReleased` event.
+        fn release(ref self: ComponentState<TContractState>, token: ContractAddress) -> u256 {
+            let amount = self.releasable(token);
+            self.Vesting_released.write(token, self.Vesting_released.read(token) + amount);
+
+            let beneficiary = get_dep_component!(@self, Ownable).owner();
+            ERC20Utils::transfer(token, beneficiary, amount);
+            self.emit(AmountReleased { token, amount });
```

Comment:
Yes, I've also noticed that the Solidity implementation emits an event prior to transferring funds. At the same time, I doubt that there's a particular sequence of events we should stick to. With that in mind I decided to follow a standard approach throughout our codebase when an event is emitted at the end of the function or close to it

Reply:
I take back my noncommittal response. The sequential order should be `AmountReleased` and then `Transfer` because releasing the funds should precede them being transferred. Do you disagree?

Reply:
Now I have to agree with you, it makes total sense from that perspective :) Changed the order of the events